### PR TITLE
Add basic routes

### PR DESCRIPTION
**Pull request by jamarob**

### CODE REVIEW

File: src/main/java/com/jamarob/jotterbackend/Note.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+import java.util.UUID;
+
+@Document
+public class Note {
+
+    @Id
+    private String id;
+    private String text;
+    private long created;
```
Comment (onepiers) [inline src/main/java/com/jamarob/jotterbackend/Note.java]: Suggestion: use an Instant

File: src/main/java/com/jamarob/jotterbackend/Note.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    private long edited;
+
+    public Note(){
+        this.id = UUID.randomUUID().toString();
```
Comment (onepiers) [inline src/main/java/com/jamarob/jotterbackend/Note.java]: MongoDb handles that for you with the @Id annotation

File: src/main/java/com/jamarob/jotterbackend/Note.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+    public String getText() {
+        return text;
+    }
+
+    public void setText(String text) {
+        this.text = text;
+    }
+
+    public long getCreated() {
```
Comment (onepiers) [inline src/main/java/com/jamarob/jotterbackend/Note.java]: Have a look at: https://codeboje.de/spring-data-auditing/

File: src/main/java/com/jamarob/jotterbackend/NoteController.java [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+@RestController
+public class NoteController {
+
+    private NoteRepository repository;
+
+    public NoteController(NoteRepository repository){
+        this.repository = repository;
+    }
+
+    @GetMapping("/ping")
```
Comment (onepiers) [inline src/main/java/com/jamarob/jotterbackend/NoteController.java]: remove or move to own controller

File: src/main/java/com/jamarob/jotterbackend/NoteController.java [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+
+    @PostMapping("/notes")
+    public NoteResponseDTO createNote(@RequestBody @Valid NoteRequestDTO dto){
+        Note note = new Note();
+        note.setText(dto.getText());
+        long now = System.currentTimeMillis();
+        note.setCreated(now);
+        note.setEdited(now);
+        repository.save(note);
+        return new NoteResponseDTO((note));
```
Comment (onepiers) [inline src/main/java/com/jamarob/jotterbackend/NoteController.java]: use the value returned from save()

File: src/main/java/com/jamarob/jotterbackend/NoteController.java [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+    }
+
+    @GetMapping("/note/{id}")
+    public NoteResponseDTO getNote(@PathVariable String id){
+        Note note = this.repository.findById(id).orElseThrow(RuntimeException::new);
```
Comment (onepiers) [inline src/main/java/com/jamarob/jotterbackend/NoteController.java]: introduce an own exception would be better -> follow up ticket

File: src/main/java/com/jamarob/jotterbackend/NoteController.java [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+    }
+
+    @PutMapping("/note/{id}")
+    public NoteResponseDTO updateNote(@PathVariable String id, @RequestBody @Valid NoteRequestDTO dto){
+        Note note = this.repository.findById(id).orElseThrow(RuntimeException::new);
+        note.setText(dto.getText());
+        note.setCreated((dto.getCreated()));
+        note.setEdited(dto.getEdited());
+        this.repository.save(note);
+        return new NoteResponseDTO(note);
```
Comment (onepiers) [inline src/main/java/com/jamarob/jotterbackend/NoteController.java]: use the return value from save()

File: src/main/java/com/jamarob/jotterbackend/NoteController.java [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+    @GetMapping("/notes")
+    public List<NoteResponseDTO> getNotes(){
+        return this.repository
+                .findAll()
+                .stream()
+                .map(NoteResponseDTO::new)
+                .collect(Collectors.toList());
+    }
+
+    @PutMapping("/notes")
```
Comment (onepiers) [inline src/main/java/com/jamarob/jotterbackend/NoteController.java]: Do you really need an saveAll endpoint()?

File: src/main/java/com/jamarob/jotterbackend/NoteController.java [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+        note.setEdited(dto.getEdited());
+        this.repository.save(note);
+        return new NoteResponseDTO(note);
+    }
+
+    @DeleteMapping("/note/{id}")
+    public NoteResponseDTO deleteNote(@PathVariable String id){
+        Note note = this.repository.findById(id).orElseThrow(RuntimeException::new);
+        this.repository.delete(note);
+        return new NoteResponseDTO(note);
```
Comment (onepiers) [inline src/main/java/com/jamarob/jotterbackend/NoteController.java]: Delete mapping should just return 202 ACCEPTED -> make it a void method

File: src/main/java/com/jamarob/jotterbackend/NoteController.java (hunk 6 above)
Comment (onepiers) [reply src/main/java/com/jamarob/jotterbackend/NoteController.java]: You actually have it: NoteNotFoundException

File: src/main/java/com/jamarob/jotterbackend/NoteController.java (hunk 4 above)
Comment (jamarob) [reply src/main/java/com/jamarob/jotterbackend/NoteController.java]: I will remove it. Was just a quick way to check if the server is alive, but I can use the getNotes() route instead.

File: src/main/java/com/jamarob/jotterbackend/NoteController.java (hunk 6 above)
Comment (jamarob) [reply src/main/java/com/jamarob/jotterbackend/NoteController.java]: Forgot to put that in.

File: src/main/java/com/jamarob/jotterbackend/NoteController.java (hunk 8 above)
Comment (jamarob) [reply src/main/java/com/jamarob/jotterbackend/NoteController.java]: I want to support undo operations and that was the easiest way I could think off in rest terms. Is there a better way?

File: src/main/java/com/jamarob/jotterbackend/Note.java (hunk 1 above)
Comment (jamarob) [reply src/main/java/com/jamarob/jotterbackend/Note.java]: Will do!

File: src/main/java/com/jamarob/jotterbackend/Note.java (hunk 3 above)
Comment (jamarob) [reply src/main/java/com/jamarob/jotterbackend/Note.java]: I will investigate that. Thanks for the link!

File: src/main/java/com/jamarob/jotterbackend/NoteController.java (hunk 7 above)
Comment (jamarob) [reply src/main/java/com/jamarob/jotterbackend/NoteController.java]: I guess its the same for the put all route?

**Comment by onepiers** [comment]
yes

Viele Grüße,
Piers

**Comment by jamarob** [comment]
I followed your advice and incorporated the changes. Thanks!
Care to take another look?

File: src/main/java/com/jamarob/jotterbackend/NoteController.java (hunk 8 above)
Comment (jamarob) [reply src/main/java/com/jamarob/jotterbackend/NoteController.java]: Nah.. you can never put in too much effort. I already have an idea how to handle it and will try to implement it in a future ticket.

**Pull request merged**